integration-testing/src/tools/network.py

```diff
@@ -0,0 +1,40 @@
+from contextlib import contextmanager
+from tools.wait import wait_for, has_peers, node_started
+import logging
+from tools.rnode import create_peer_nodes
+
+
+class RChain:
+    def __init__(self, network, bootstrap, peers):
+        self.network = network
+        self.bootstrap = bootstrap
+        self.peers = peers
+        self.nodes = [bootstrap] + peers
+
+
+@contextmanager
+def start_network(config, docker, bootstrap, validators_data, allowed_peers=None):
+    logging.debug(f"Docker network = {bootstrap.network}")
+
+    bonds_file, _, peer_keys = validators_data
+    peers = create_peer_nodes(docker, bootstrap, bootstrap.network, bonds_file, peer_keys, config.rnode_timeout, allowed_peers)
+
+    yield RChain(network = bootstrap.network, bootstrap = bootstrap, peers = peers)
+
+    for peer in peers:
+        peer.cleanup()
+
+
+def wait_for_started_network(node_startup_timeout, network):
+    for peer in network.peers:
+        wait_for(node_started(peer), node_startup_timeout, f"Peer {peer.name} did not start correctly.")
+
+def wait_for_converged_network(timeout, network, peer_connections):
+    wait_for( has_peers( network.bootstrap, len(network.peers)),
```

Comment:
Ditto, see the callbacks argument

Reply:
I don't understand the problem. Without a basic tool like wait_for and its predicates you will end up with tons of functions wait_for_node, wait_for_peers, wait_for_lunch...

Maybe I should document wait_for and its mechanisms better?

Reply:
> Without a basic tool like wait_for and its predicates you will end up with tons of functions wait_for_node, wait_for_peers, wait_for_lunch...

This is exactly what I'm proposing.  The savings in repeated code nowhere near offset the detriment to readability and simplicity.

Reply:
I was focusing on readability from the final user perspective, where the final user is the Scala developer, with minimal Python knowledge, who wants to write an intregration test. With these tools the code that waits for the bootstrap node to start is this:

```python
def start_bootstrap(docker_client, node_start_timeout, node_cmd_timeout, validators_data):
    with docker_network(docker_client) as network:
        with create_bootstrap(docker_client, network, node_cmd_timeout, validators_data) as node:

            wait_for( node_started(node),
                      node_start_timeout,
                      "Bootstrap node didn't start correctly")

            yield node
```

...and a test failure because the bootstrap node failed to start would look like this:

```
docker_client = <docker.client.DockerClient object at 0x10445dd30>, node_start_timeout = 50, node_cmd_timeout = 10
validators_data = ValidatorsData(bonds_file='/tmp/rchain-bonds-file-zdyeeawm.txt', bootstrap_keys=KeyPair(private_key='2574c8e500a0de1d3...9455377bf8a805d54dd3933612aeb3d970a5', public_key='8b9f35f7cb951b4c89d450e418c921e6faf484f93c94a04ec00e310b624339d0')])

    @contextmanager
    def start_bootstrap(docker_client, node_start_timeout, node_cmd_timeout, validators_data):
        with docker_network(docker_client) as network:
            with create_bootstrap(docker_client, network, node_cmd_timeout, validators_data) as node:

                wait_for( node_started(node),
                          node_start_timeout,
>                         "Bootstrap node didn't start correctly")
E               Failed: Bootstrap node didn't start correctly

src/tools/rnode.py:193: Failed
```

I really think this is as simple and readable as you can probably get.

Reply:
Both variants are equally readable to me.

OTOH I think - if not for other concerns - all of us would prefer a composable, no-duplication apporach like the higher-order functions one. 

All that being said, my feeling is that after seeing a stacktrace for a failed `wait_for` we'll all be on the same page.

@odeac can you provide an example stacktrace when this fails due to 1) timeout after many checks, 2) failed requirement (not timeout)?

@adaszko can you provide yours as well, so that we're sure the issues you've mentioned are present in the example?

Reply:
This shows how a failure looks with such a tool: https://gist.github.com/odeac/171667c9bee66a317dd8c22a678572af

For me it's a no brainer that in this case we should prefer a solution with focus on readability .

Reply:
It would be nice if a message such as `Condition not fulfilled yet (same as above). Time left: 20s. Sleeping 3s...` said *what* condition it is waiting to be satisfied.  The word 'Condition' by itself isn't helpful to the user.

Reply:
Anyway, I vote with @odeac here. I don't see why you wouldn't use `wait_for` since it's a standard part of the coroutines library. If it becomes a problem to debug, @adaszko will get to say, "I told you so!" But, I'm dubious since the output @odeac provided seems pretty readable and the code is definitely nice and readable.

Reply:
> It would be nice if a message such as `Condition not fulfilled yet (same as above). Time left: 20s. Sleeping 3s...` said _what_ condition it is waiting to be satisfied. The word 'Condition' by itself isn't helpful to the user.

I choose to print "same as above" for two reasons:
1. to reduce clutter in the logs
2. to make it obvious that the status hasn't changed since the previous check. Otherwise, if the difference in the messages is subtle, the difference could be easy to miss